### PR TITLE
Add borrowing between priority levels in APF

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
This PR implement the design change in https://github.com/kubernetes/enhancements/pull/3391 and https://github.com/kubernetes/enhancements/pull/3479 .  That is, this PR adds borrowing of concurrency between priority levels in the API Priority and Fairness feature.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
The API change part of this was already reviewed in #112830 .
The first six commits of this PR are the commits of that earlier PR, rebased onto a later revision of the `master` branch.
The "apiserver: update borrowing parameters for apf bootstrap objects" commit is a cherry-pick of #113016 .

#### Does this PR introduce a user-facing change?

```release-note
Priority and Fairness has introduced a new feature called _borrowing_ that allows an API priority level
to borrow a number of seats from other priority level(s). As a cluster operator, you can enable borrowing
for a certain priority level configuration object via the two newly introduced fields `lendablePercent`, and
`borrowingLimitPercent` located under the `.spec.limited` field of the designated priority level.
This PR adds the following metrics.
- `apiserver_flowcontrol_nominal_limit_seats`: Nominal number of execution seats configured for each priority level
- `apiserver_flowcontrol_lower_limit_seats`: Configured lower bound on number of execution seats available to each priority level
- `apiserver_flowcontrol_upper_limit_seats`: Configured upper bound on number of execution seats available to each priority level
- `apiserver_flowcontrol_demand_seats`: Observations, at the end of every nanosecond, of (the number of seats each priority level could use) / (nominal number of seats for that level)
- `apiserver_flowcontrol_demand_seats_high_watermark`: High watermark, over last adjustment period, of demand_seats
- `apiserver_flowcontrol_demand_seats_average`: Time-weighted average, over last adjustment period, of demand_seats
- `apiserver_flowcontrol_demand_seats_stdev`: Time-weighted standard deviation, over last adjustment period, of demand_seats
- `apiserver_flowcontrol_demand_seats_smoothed`: Smoothed seat demands
- `apiserver_flowcontrol_target_seats`: Seat allocation targets
- `apiserver_flowcontrol_seat_fair_frac`: Fair fraction of server's concurrency to allocate to each priority level that can use it
- `apiserver_flowcontrol_current_limit_seats`: current derived number of execution seats available to each priority level

The possibility of borrowing means that the old metric apiserver_flowcontrol_request_concurrency_limit can no longer mean both the configured concurrency limit and the enforced concurrency limit.  Henceforth it means the configured concurrency limit.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
[KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1040-priority-and-fairness#dispatching

```

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
/cc @tkashem 
/cc @wojtek-t 
@cyang49

**Comment by MikeSpreitzer** [comment]
@lavalamp

**Comment by lavalamp** [comment]
I'm reading the last few commits, then

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc_test.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+					lowSum += classes[j].lowerBound
+					highSum += classes[j].upperBound
+				})
+			default:
+				style = "general"
+				for j := 0; j < probLen; j++ {
+					x := f32max(0, rands.Float32()*5-1)
+					classes[j].lowerBound = x
+					classes[j].target = x + 2*rands.Float32()
+					classes[j].upperBound = x + 3*rands.Float32()
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc_test.go:66]: This could produce a target greater than the upperBound?

**Review by lavalamp — COMMENTED** [review]
I haven't yet reviewed the last commit, but some comments on the three before that.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go [hunk 2]
```diff
@@ -316,6 +317,125 @@ var (
...
+			StabilityLevel: compbasemetrics.ALPHA,
+		},
+		[]string{priorityLevel},
+	)
+	ApiserverSeatDemands = NewTimingRatioHistogramVec(
+		&compbasemetrics.TimingHistogramOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "demand_seats",
+			Help:           "Observations, at the end of every nanosecond, of (the number of seats each priority level could use) / (nominal number of seats for that level)",
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go:345]: The borrowing design makes periodic adjustments based on the pressure in each priority level, and this metric reflects that pressure, so it needs to be per priority level.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go [hunk 3]
```diff
@@ -316,6 +317,125 @@ var (
...
+			Buckets:        []float64{0.2, 0.4, 0.6, 0.8, 1, 1.2, 1.4, 1.7, 2, 2.8, 4, 6},
+			StabilityLevel: compbasemetrics.ALPHA,
+		},
+		priorityLevel,
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go:349]: `NewTimingRatioHistogramVec` takes a variadic parameter where others take a slice, making the calls more succinct.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/integrator.go [hunk 4]
```diff
@@ -77,6 +78,24 @@ func (igr *integrator) Set(x float64) {
 	igr.Unlock()
 }
 
+func (igr *integrator) Add(deltaX float64) {
+	igr.Lock()
+	igr.setLocked(igr.x + deltaX)
+	igr.Unlock()
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/integrator.go:84]: If I understand correctly it does not matter in this case because `setLocked` cannot panic.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go [hunk 5]
```diff
@@ -177,6 +177,9 @@ type configController struct {
 	// name to the state for that level.  Every name referenced from a
 	// member of `flowSchemas` has an entry here.
 	priorityLevelStates map[string]*priorityLevelState
+
+	// nominalCLSum is the sum of the nominalCL fields in the priorityLevelState records
+	nominalCLSum int
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:182]: See https://github.com/kubernetes/enhancements/pull/3643 for why the server concurrency limit is not necessarily the required sum going into the concurrency solver.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+		return nil, 0, errors.New("negative sums are not supported")
+	}
+	requiredSumF := float32(requiredSum)
+	var lowSum, highSum, targetSum float32
+	ubRange := minMax{min: float32(math.MaxFloat32)}
+	lbRange := minMax{min: float32(math.MaxFloat32)}
+	relativeItems := make([]relativeAllocItem, len(classes))
+	for idx, item := range classes {
+		target := item.target
+		if target < item.lowerBound {
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go:126]: The target can be greater than the upper bound.  The KEP explains the reason for the asymmetric imposition of bounds on the target.  In https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1040-priority-and-fairness#dispatching look for the following.

> This shares the wealth or the pain proportionally among the priority levels (but note: the upper bound does not affect the target, lest the pain of not achieving a high SmoothSeatDemand be distorted, while the lower bound does affect the target, so that merely achieving the lower bound is not considered a gain).

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+			return nil, 0, fmt.Errorf("target %d is %v, which is below its lower bound of %v", idx, target, item.lowerBound)
+		}
+		if item.upperBound <= 0 {
+			return nil, 0, fmt.Errorf("upper bound %d is %v but should be positive", idx, item.upperBound)
+		}
+		if target == 0 {
+			// tweak this to a non-zero value so avoid dividing by zero
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go:133]: `target` is not a bound.
The point here is just to avoid dividing by zero.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc_test.go (hunk 1 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc_test.go:66]: Yes, that is an allowed case.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+// if possible otherwise returns an error saying why it is impossible.
+// `allocs` sums to `requiredSum`.
+// For each J in [0, len(classes)):
+// (1) `classes[J].lowerBound <= allocs[J] <= classes[J].upperBound` and
+// (2) exactly one of the following is true:
+// (2a) `allocs[J] == fairProp * classes[J].target`,
+// (2b) `allocs[J] == classes[J].lowerBound && classes[J].lowerBound > fairProp * classes[J].target`, or
+// (2c) `allocs[J] == classes[J].upperBound && classes[J].upperBound < fairProp * classes[J].target`.
+// Each allocProblemItem is required to have `target >= lowerBound` and `upperBound > 0`.
+// A target of zero is treated as if it were 0.001.
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go]: Done in the latest revision.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+	var lowSum, highSum, targetSum float32
+	ubRange := minMax{min: float32(math.MaxFloat32)}
+	lbRange := minMax{min: float32(math.MaxFloat32)}
+	relativeItems := make([]relativeAllocItem, len(classes))
+	for idx, item := range classes {
+		target := item.target
+		if target < item.lowerBound {
+			return nil, 0, fmt.Errorf("target %d is %v, which is below its lower bound of %v", idx, target, item.lowerBound)
+		}
+		if item.upperBound <= 0 {
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go]: Done in the latest revision.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+			relativeUpperBound: item.upperBound / target,
+		}
+		ubRange.note(relativeItem.relativeUpperBound)
+		lbRange.note(relativeItem.relativeLowerBound)
+		relativeItems[idx] = relativeItem
+	}
+	if lbRange.max > 1 {
+		panic(lbRange)
+	}
+	if lowSum-requiredSumF > 0.00001 {
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+		return nil, 0, fmt.Errorf("lower bounds sum to %v, which is higher than the required sum of %v", lowSum, requiredSum)
+	}
+	if requiredSumF-highSum > 0.00001 {
+		return nil, 0, fmt.Errorf("upper bounds sum to %v, which is lower than the required sum of %v", highSum, requiredSum)
+	}
+	ans := make([]float32, len(classes))
+	if requiredSum == 0 {
+		return ans, 0, nil
+	}
+	if lowSum-requiredSumF > -0.001 { // no wiggle room, constrained from below
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go]: Simplified in the latest revision, this is really only about floating-point fuzz.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+		}
+		return ans, fairProp, nil
+	}
+	// Sadly, some bounds matter.
+	// We find the solution by sorting the bounds and considering progressively
+	// higher values of fairProp, starting from lbRange.min.
+	rap := (&relativeAllocProblem{items: relativeItems}).initIndices()
+	sumSoFar := lowSum
+	fairProp = lbRange.min
+	var weight, dWeight float32
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go]: "delta".  Spelled out in the latest revision.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+	var weighing, dWeighing int
+	lastIdx := -1
+	// `lastIdx` is the last `rap` index that has been considered.
+	// If the current value of fairProp were the answer then
+	// `sumSoFar == requiredSum`.
+	// Otherwise the next increment in fairProp involves changing the allocations
+	// of `weighing` classes whose targets sum to `weight`.
+	// Raising fairProp above `lastidx` adds `dWeight` to `weight`
+	// and `dWeighing` to `weithing`.
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go]: Fixed, thanks.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go (hunk 5 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:185]: Latest revision expands the commentary here.

File: staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go [hunk 14]
```diff
@@ -96,6 +97,7 @@ var (
 			Type: flowcontrol.PriorityLevelEnablementLimited,
 			Limited: &flowcontrol.LimitedPriorityLevelConfiguration{
 				NominalConcurrencyShares: 5,
+				LendablePercent:          pointer.Int32(0),
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go:100]: I was trying to figure out if it would have been safe for mandatory priority levels to use a non-default value for this field, since API validation compares the incoming object's `spec` with the internal default's `spec` and fails if there is any difference:

https://github.com/kubernetes/kubernetes/blob/548d135ea6b3b9a68d31c981b65e1f9e9643ea58/pkg/apis/flowcontrol/validation/validation.go#L343-L355

In a skewed API server scenario, an update to a newer object containing the field could be sent to an older server, and vice versa.

<=1.25 servers would be unaware of this field and would just drop it.

>=1.26 servers would default a missing value for the field to the default value. That happens to match what is specified here, but if this wanted to use a non-default value (like some of the suggested priority levels below), I think update validation would fail.

If it is the case that it is only safe to use default lendablePercent values for the mandatory priority level configurations (at least until v1beta3+ is the oldest served version in 1.29+, and we know all clients are aware of the field), can we record that with a comment so future adders/editors don't introduce a non-default lendablePercent value without realizing the consequences?

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go:100]: I see that https://kubernetes.io/docs/concepts/cluster-administration/flow-control/#mandatory-configuration-objects does not state this clearly enough: it is not intended that any users, not even cluster admins, modify the specs of the mandatory objects.

The config-providing controller in each apiserver talks specifically to that apiserver.

In a mixed-release scenario there may be a continual fight between apiservers of different releases.  We have looked for better alternatives, but not found one.  We expect a mixture of releases in one cluster to persist for a limited amount of time --- the duration of a rollout or roll-back.

**Comment by MikeSpreitzer** [comment]
The force-push to 66b99cf804c is a rebase onto `master` to pick up the related PRs that just merged.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+			reqInfo := &request.RequestInfo{
+				IsResourceRequest: false,
+				Path:              "/foobar",
+				Verb:              "GET",
+			}
+			noteFn := func(fs *flowcontrol.FlowSchema, plc *flowcontrol.PriorityLevelConfiguration, fd string) {}
+			workEstr := func() fcrequest.WorkEstimate { return fcrequest.WorkEstimate{InitialSeats: 1} }
+			qnf := fq.QueueNoteFn(func(bool) {})
+			var wg sync.WaitGroup
+			wg.Add(30)
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go:184]: Did you mean to also have a `defer wg.Done()` here or something like that?

Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go:184]: I think I have read the whole thing now and this is the only thing I saw.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go [hunk 16]
```diff
@@ -0,0 +1,252 @@
+/*
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go:1]: I added couple comments about the solver into: https://github.com/kubernetes/kubernetes/pull/112137#pullrequestreview-1170103370

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go [hunk 17]
```diff
@@ -221,16 +221,6 @@ var (
 		},
 		[]string{priorityLevel, flowSchema},
 	)
-	apiserverRequestConcurrencyLimit = compbasemetrics.NewGaugeVec(
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go:224]: Let's not delete the metric in this large PR - please move it to a separate PR (with explicit approval from sig instrumentation folks).

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go:224]: The problem here is that this metric means two things (NominalCL, CurrentCL) that are equal before this PR but not generally equal after this PR.  Keeping it would be, at best, misleading.  And which meaning would be best to keep for this metric?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go (hunk 15 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go:184]: The corresponding `wg.Done()` is inside the goroutine inside the loop below, at line 204.

Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go:184]: Sorry, I meant to type `Wait()`

Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go:184]: (Add and Done are called, but AFAICT not Wait, which makes it pointless)

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go:184]: The force-push to 7c06996bb9d moves that `wg.Done()` to the top of the goroutine, with a `defer`.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go (hunk 16 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go:1]: The force-push to 7c06996bb9d picks up my response to that review.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go (hunk 15 above)
Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go:184]: Still need to call `wg.Wait()` somewhere?

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go:184]: D'Oh!

The previous force-push did not actually change the Done, I forgot to commit.
The force-push to a36f07e6ad6 changes the Done and adds the Wait.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+				}
+				nThreads := 24 - flow*18
+				for thread := 0; thread < nThreads; thread++ {
+					go func() {
+						wg.Done()
+						wait.Until(func() { controller.Handle(ctx, rd, noteFn, workEstr, qnf, exec) }, 0, ctx.Done())
+					}()
+				}
+			}
+			wg.Wait()
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go:208]: I was expecting to see this at the end or via defer, I thought the threads would run until the `cancel()` on line 221 is executed?

Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go:208]: Wait, the wg is to make sure the thread has started, not to make sure it has ended?

Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go:208]: Considering the Sleep(10) and the fact that everything passed prior to adding this call, I'm not sure this waitgroup is having an effect on the test.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go (hunk 15 above)
Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go:184]: Now that I look at it again, this calls Add(30) but I had to do math to ensure that Done is actually called that many times. Ordinarily I'd ask to put Add(1) calls prior to starting the goroutines, but see my next comment because I'm not convinced that this waitgroup is actually needed, can you just remove it? Or modify to use it to make sure all the goroutines have exited before letting the test end (this is what I assumed it was for originally).

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go (hunk 17 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go:224]: @logicalhan , @dgrisonnet : any thoughts here?

Comment (dgrisonnet) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go:224]: I agree with @wojtek-t, removing the metric in a separate would make it easier to track and document in a release note.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go (hunk 16 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go:1]: That PR has been already merged.
@MikeSpreitzer - can you please rebase?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go [hunk 19]
```diff
@@ -225,6 +231,18 @@ type priorityLevelState struct {
 	// Periodically smoothed gets replaced with `max(envelope, A*smoothed + (1-A)*envelope)`,
 	// where A is seatDemandSmoothingCoefficient.
 	seatDemandStats seatDemandStats
+
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:234]: @MikeSpreitzer - can you please squash the last two commits - they are tightly coupled with each other (first doesn't make sense without the second).

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go [hunk 20]
```diff
@@ -368,12 +377,69 @@ func (cfgCtlr *configController) Run(stopCh <-chan struct{}) error {
...
+	plNames := make([]string, 0, len(plStates))
+	for plName, plState := range plStates {
+		if plState.pl.Spec.Limited == nil {
+			continue
+		}
 		obs := plState.seatDemandIntegrator.Reset()
 		plState.seatDemandStats.update(obs)
-		metrics.NotePriorityLevelConcurrencyAdjustment(plState.pl.Name, plState.seatDemandStats.highWatermark, plState.seatDemandStats.avg, plState.seatDemandStats.stdDev, plState.seatDemandStats.smoothed /* TODO: add the designed rest for borrowing */)
-		// TODO: updathe CurrentCL as described in https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1040-priority-and-fairness#dispatching
+		minCurrentCL := math.Max(float64(plState.minCL), math.Min(float64(plState.nominalCL), plState.seatDemandStats.highWatermark))
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:392]: nit: I think I figured out where this formula is coming from, but can you please add a brief comment explaining it?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go [hunk 21]
```diff
@@ -0,0 +1,250 @@
...
+				concIntegrators[flow] = concIntegrator
+				exec := func() {
+					concIntegrator.Inc()
+					clk.Sleep(250 * time.Millisecond)
+					concIntegrator.Dec()
+				}
+				nThreads := 24 - flow*18
+				for thread := 0; thread < nThreads; thread++ {
+					go func() {
+						wg.Done()
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go:203]: shouldn't this be
`defer wg.Done()`

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go:203]: ohh - I see Daniel`s comment below - it touches the same thing I guess.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go (hunk 17 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go:224]: >  And which meaning would be best to keep for this metric?

Let's open the PR removing it now and merge that right after this PR merges.
If we do that - we can choose whatever meaning you prefer.

**Comment by sftim** [comment]
Release note suggestion
```markdown
Priority and Fairness has introduced a new feature called _borrowing_ that allows an API priority level
to borrow a number of seats from other priority level(s). As a cluster operator, you can enable borrowing
for a certain priority level configuration object via the two newly introduced fields `lendablePercent`, and
`borrowingLimitPercent` located under the `.spec.limited` field of the designated priority level.
```

I recommend adding a heading to the KEP document, “Borrowing”. It's OK to borrow the text about that from the end-user docs that we write for the new behavior.

File: staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go (hunk 14 above)
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap/default.go:100]: I would expect the bootstrap configuration to use loopback connections, which point back to *this* kube-apiserver, not to the general HA endpoint, so during bootstrapping, the new kube-apiserver will only send bootstrap requests to itself and the old kube-apiserver will only send bootstrap requests to itself.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go (hunk 17 above)
Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go:224]: I'm okay with either decision.

**Comment by lavalamp** [comment]
I'm gonna tag this to make sure it meets the code freeze criteria

/lgtm
/approve

@MikeSpreitzer if you could rebase that would be great, I can live with it if you fix the nits in a followup.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go (hunk 17 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go:224]: No!  The problem occurs right here in this PR.  This PR make a priority level's configured concurrency limit not necessarily  the same as the concurrency limit that is enforced.

Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/metrics.go:224]: Can you amend the release note to include the removal of this metric?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go (hunk 18 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go:208]: Yes, that wait group is to make sure the goroutines have started.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go (hunk 19 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:234]: Done in the force-push to 722c85844e3

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go (hunk 20 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:397]: Done.

**Comment by lavalamp** [comment]
/lgtm

Can you update the release note text per https://github.com/kubernetes/kubernetes/pull/113485#discussion_r1017137702?

File: pkg/apis/flowcontrol/internalbootstrap/defaults_test.go [hunk 22]
```diff
@@ -17,31 +17,45 @@ limitations under the License.
 package internalbootstrap
 
 import (
+	"fmt"
 	"testing"
 
+	"github.com/google/go-cmp/cmp"
 	flowcontrol "k8s.io/api/flowcontrol/v1beta3"
```
Comment (logicalhan) [inline pkg/apis/flowcontrol/internalbootstrap/defaults_test.go:24]:
```suggestion
	"github.com/google/go-cmp/cmp"

	flowcontrol "k8s.io/api/flowcontrol/v1beta3"
```

File: pkg/apis/flowcontrol/v1alpha1/defaults_test.go [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+*/
+
+package v1alpha1
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	flowcontrolv1alpha1 "k8s.io/api/flowcontrol/v1alpha1"
```
Comment (logicalhan) [inline pkg/apis/flowcontrol/v1alpha1/defaults_test.go:24]:
```suggestion
	"github.com/google/go-cmp/cmp"

	flowcontrolv1alpha1 "k8s.io/api/flowcontrol/v1alpha1"
```

File: pkg/apis/flowcontrol/v1beta1/defaults_test.go [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+*/
+
+package v1beta1
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	flowcontrolv1beta1 "k8s.io/api/flowcontrol/v1beta1"
```
Comment (logicalhan) [inline pkg/apis/flowcontrol/v1beta1/defaults_test.go:24]:
```suggestion
	"github.com/google/go-cmp/cmp"

	flowcontrolv1beta1 "k8s.io/api/flowcontrol/v1beta1"
```

File: pkg/apis/flowcontrol/v1beta2/defaults_test.go [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+*/
+
+package v1beta2
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	flowcontrolv1beta2 "k8s.io/api/flowcontrol/v1beta2"
```
Comment (logicalhan) [inline pkg/apis/flowcontrol/v1beta2/defaults_test.go:24]:
```suggestion
	"github.com/google/go-cmp/cmp"

	flowcontrolv1beta2 "k8s.io/api/flowcontrol/v1beta2"
```

File: pkg/apis/flowcontrol/v1beta3/defaults_test.go [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+*/
+
+package v1beta3
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	flowcontrolv1beta3 "k8s.io/api/flowcontrol/v1beta3"
```
Comment (logicalhan) [inline pkg/apis/flowcontrol/v1beta3/defaults_test.go:24]:
```suggestion
	"github.com/google/go-cmp/cmp"

	flowcontrolv1beta3 "k8s.io/api/flowcontrol/v1beta3"
```

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go (hunk 16 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go]: done

**Comment by MikeSpreitzer** [comment]
The force-push to feb42277884 renames the wait group to make its purpose clearer and switches from a formula to an array for the number of clients per flow.

**Comment by MikeSpreitzer** [comment]
I updated the release note to list the new metrics and explain what happened to `apiserver_flowcontrol_request_concurrency_limit`.

**Comment by wojtek-t** [comment]
I'm retagging after just nits were addressed based on the fact that it was tagged before code-freeze in:
https://github.com/kubernetes/kubernetes/pull/113485#issuecomment-1307883479

FWIW - it LGTM too.

/lgtm
/approve
/milestone v1.26
